keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,216 @@
+  - [Current status of dependencies](#current-status-of-dependencies)
+  - [Current cgroups usage and the equivalent in cgroups v2](#current-cgroups-usage-and-the-equivalent-in-cgroups-v2)
+- [Proposal](#proposal)
+  - [Option 1: Convert from cgroups v1 settings to v2](#option-1-convert-from-cgroups-v1-settings-to-v2)
+  - [Option 2: Use cgroups v2 throughout the stack](#option-2-use-cgroups-v2-throughout-the-stack)
+- [Risk and Mitigations](#risk-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing](#testing)
+
+<!-- /toc -->
+
+## Summary
+
+A proposal to add support for cgroups v2 to kubernetes.
+
+## Motivation
+
+The new kernel cgroups v2 API was declared stable more than two years
+ago. Newer features in the kernel such as PSI depend upon cgroups
+v2. groups v1 will eventually become obsolete in favor of cgroups v2.
+Some distros are already using cgroups v2 by default, and that
+prevents Kubernetes from working as it is required to run with cgroups
+v1.
+
+## Goals
+
+This proposal aims to:
+
+*   Add support for cgroups v2 to cAdvisor ([https://github.com/google/cadvisor/pull/2309](https://github.com/google/cadvisor/pull/2309))
+*   Add support for cgroups v2 to the Kubelet
+
+## User Stories
+
+*   The Kubelet can run on a host using either cgroups v1 or v2
+
+## Implementation Details
+
+### Dependencies on OCI and container runtimes
+
+Cgroups v2 requires changes to the OCI runtime specs. Changes to the
+OCI specs are out of the scope for this proposal.
+
+### Current status of dependencies
+
+- CRI-O+crun: support cgroups v2
+
+- runc support for cgroups v2 is work in progress
+
+- containerd: [https://github.com/containerd/containerd/pull/3799](https://github.com/containerd/containerd/pull/3799)
+
+- Moby: [https://github.com/moby/moby/pull/40174](https://github.com/moby/moby/pull/40174)
+
+- OCI runtime spec: TODO
+
+## Current cgroups usage and the equivalent in cgroups v2
+
+|Kubernetes cgroups v1|Kubernetes cgroups v2 behavior|
+|---|---|
+|CPU stats for Horizontal Pod Autoscaling|No .percpu cpuacct stats.  Check cadvisor.|
+|CPU pinning based on integral cores|Cpuset controller available|
+|Memory limits|Not changed, different naming|
+|PIDs limits|Not changed, same naming|
+|hugetlb|Added to linux-next, targetting Linux 5.6|
+
+## Proposal
+
+The proposal is to implement cgroups v2 in two different phases.
+
+The first phase ensures that any configuration file designed for
+cgroups v1 will continue to work on cgroups v2.
+
+The second phase requires changes through the entire stack, including
+the OCI runtime specifications.
+
+At startup the Kubelet detects what hierarchy the system is using.  It
+checks the file system type for `/sys/fs/cgroup` (the equivalent of
+`stat -f --format '%T' /sys/fs/cgroup`).  If the type is `cgroup2fs`
+then the Kubelet will use only cgroups v2 during all its execution.
+
+### cgroup namespace
+
+A cgroup namespace restricts the view on the cgroups.  When
```

Comment:
it is not clear how this section maps to the kep.  it feels like it is referencing rootless requirements, which in 1/28 sig-node meeting we discussed as not being an initial target.  if you do not intend to use the cgroup namespace because the kubelets view on cgroups should not be restricted, please omit it.  cadvisor metrics endpoints can still be enabled in the kubelet so i feel the kubelet still needs to have full visibility of cgroups on host.

/cc @dashpole

Reply:
No, unrelated to rootless

Reply:
More context is available here: https://github.com/containers/libpod/issues/4363

Reply:
as Akihiro pointed out, this is not related to rootless.  It is useful to limit the container's view on cgroups.  Differently than cgroup v1, every kernel supporting cgroups v2 has support for the cgroup namespace so it makes sense to enable it by default:

```
## Without a cgroup namespace
# podman run --cgroupns=host fedora cat /proc/self/cgroup
0::/machine.slice/libpod-c8aabfba22cae68f60063d63cd6a9c17d91cf7bbf3cf9052b13d78d6af7d37b0.scope

## With a cgroup namespace
# podman run --cgroupns=private fedora cat /proc/self/cgroup
0::/
```

Reply:
@giuseppe @AkihiroSuda thanks for the clarification.

we need to be able to preserve the ability to run monitoring agents that have full visibility.  is usage of this namespace only enabled if the container is not privileged?  for example, we will soon require the need to run cAdvisor as a pod rather than via the kubelet, and that pod will need full host visibility.

Reply:
Even privileged pod should unshare cgrouns, as well as netns and pidns.

Adding new podspec field seems consistent with netns and pidns

Reply:
Can you summarize what we discussed above, what we agreed and not-agreed and unclear in the doc? For example, we are not clear on how to support privileged pod which should have the full view of the cgroups on the host.

Reply:
for privileged pods we can default to use the host cgroupns, so that we won't break existing privileged containers.

I'll specify it in the KEP

Reply:
Privileged pods are already going to be broken because of incompatibility across v1 and v2.

Why not treat cgroupns just same as pidns and netns?

Reply:
Privileged pods will be broken if the container payload doesn't handle cgroup v2, but that doesn't depend from the container configuration.

Changing the default means the configuration must be changed to maintain the current behaviour.